clippy_lints/src/if_then_some_else_none.rs

```diff
@@ -79,6 +80,7 @@ impl<'tcx> LateLintPass<'tcx> for IfThenSomeElseNone {
             && !expr.span.in_external_macro(cx.sess().source_map())
             && self.msrv.meets(cx, msrvs::BOOL_THEN)
             && !contains_return(then_block.stmts)
+            && !expr_requires_coercion(cx, then_expr)
```

Comment:
In this case, what do you think about, instead of completely discarding the block if it requires type coercion, we check if we're in a trait item function (we cannot change the signatures of those functions) and if we're in just a normal function, we emit a non-`MachineApplicable` suggestion with a note about the user will have to modify the return type?